### PR TITLE
Use fields mask

### PR DESCRIPTION
**Pull request by noirbizarre**
This PR make use of fields Mask and GenericModelFields.

Extracted from #380 (so requires #380 )

### CODE REVIEW

File: udata/commands/fixtures.py [hunk 1]
```diff
@@ -5,20 +5,19 @@
 import logging
 
 from udata.commands import manager
-from udata.core.user.factories import UserFactory, AdminFactory
-from udata.core.dataset.factories import (
-    VisibleDatasetFactory, DatasetDiscussionFactory, LicenseFactory
-)
-from udata.core.reuse.factories import VisibleReuseFactory
+from udata.core.dataset.factories import VisibleDatasetFactory, LicenseFactory
```
Comment (vinyll) [inline udata/commands/fixtures.py]: Bewarn that it may conflict with https://github.com/etalab/udata/pull/444

File: udata/tests/forms/test_reference_field.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 from udata.forms import ModelForm, fields
 from udata.models import db
 from udata.tests import TestCase
-from udata.tests.factories import faker
+from udata.core.user.factories import faker
```
Comment (vinyll) [inline udata/tests/forms/test_reference_field.py]: In https://github.com/etalab/udata/pull/444/commits/1b3226924f7aa46608a6e56aae04d0de4532e91f#diff-718f5796155236b92d81115cf73d992cR208 faker was moved to `udata.utils`

File: udata/features/oembed/api.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 from udata import theme
 from udata.api import api, API, output_json
-from udata.models import db, Dataset, GeoZone
+from udata.models import db, Dataset, GeoZone, TERRITORY_DATASETS
```
Comment (vinyll) [inline udata/features/oembed/api.py]: I guess with should move this out from _udata_ to _datagouv_

File: js/components/discussions/list.vue [hunk 4]
```diff
@@ -37,10 +37,10 @@ export default {
     },
     events: {
         'datatable:item:click': function(discussion) {
-            let prefix = discussion.class.replace('Discussion', '').toLowerCase(),
+            let prefix = discussion.subject.class.toLowerCase(),
```
Comment (vinyll) [inline js/components/discussions/list.vue]: `let` vs `const` (eg: https://github.com/etalab/udata/pull/451/commits/45501bf92e4542163142d50664826b5daeec1941#diff-c0bbbc951175bcb27057d4ec96470f3eR46)

File: udata/core/organization/api.py [hunk 5]
```diff
@@ -87,8 +87,13 @@ def get(self, org):
...
+    @api.response(410, 'Organization has been deleted')
     def put(self, org):
-        '''Update a organization given its identifier'''
+        '''
+        Update a organization given its identifier
+
+        :raises PermissionDenied:
+        '''
         if org.deleted:
             api.abort(410, 'Organization has been deleted')
```
Comment (vinyll) [inline udata/core/organization/api.py]: Is this still needed with the decorator _response 410_ ?

Comment (vinyll) [inline udata/core/organization/api.py]: should https://github.com/etalab/udata/pull/451/commits/45501bf92e4542163142d50664826b5daeec1941#diff-6ce9a6ba822a15f73a6b9269b1a7c62aR109 also get a decorator?

File: udata/core/user/api.py [hunk 6]
```diff
@@ -205,25 +210,67 @@ def delete(self):
...
-    model = User
-    fields = user_fields
-    form = UserProfileForm
+@api.response(404, 'User not found')
+@api.response(410, 'User has been deleted')
+class UserAPI(API):
+    @api.doc('get_user')
+    @api.marshal_with(user_fields)
+    def get(self, user):
+        '''Get an user given its identifier'''
```
Comment (vinyll) [inline udata/core/user/api.py]: _Delete a user_ (_u_ being pronounced 'you' there's no 'n')
This applies for other '_an user_' comments in this PR

File: udata/templates/macros/issues.html [hunk 7]
```diff
@@ -1,8 +1,8 @@
 {% macro issues_btn(obj, icons=None) %}
-{% set basename = obj.__class__.__name__|lower %}
 <button type="button" class="btn btn-danger btn-issues"
     data-api-url="{{ url_for('api.issues', for=obj.id|string) }}"
-    data-subject="{{ obj.id }}"
+    data-subject-id="{{ obj.id }}"
+    data-subject-class="{{ obj.__class__.__name__ }}"
```
Comment (vinyll) [inline udata/templates/macros/issues.html]: shouldn't we have a property in the obj (like `class_name`?) that returns `self.__class__.__name__` and use that instead here?

File: udata/tests/test_issues.py [hunk 8]
```diff
@@ -178,6 +178,36 @@ def test_list_issues_closed_filter(self):
...
+            issues.append(issue)
+        # Creating a closed issue that shouldn't show up in response.
+        user = UserFactory()
+        other_dataset = Dataset.objects.create(title='Other Test dataset')
+        message = Message(content=faker.sentence(), posted_by=user)
+        issue = Issue.objects.create(
+            subject=other_dataset,
+            user=user,
+            title='test issue {}'.format(i),
+            discussion=[message]
```
Comment (vinyll) [inline udata/tests/test_issues.py]: I had never noticed this _discussion_ property name was in singular form.
All other _db.ListField_ properties in the project have a name in the plural form.

File: udata/commands/fixtures.py (hunk 1 above)
Comment (noirbizarre) [inline udata/commands/fixtures.py]: Already rebased and fixed in last commit ;)

File: udata/features/oembed/api.py (hunk 3 above)
Comment (noirbizarre) [inline udata/features/oembed/api.py]: Nop, it's only populated with french territories in gouvfr

File: js/components/discussions/list.vue (hunk 4 above)
Comment (noirbizarre) [inline js/components/discussions/list.vue]: noted :+1:

File: udata/core/organization/api.py (hunk 5 above)
Comment (noirbizarre) [inline udata/core/organization/api.py]: The decorator is only here to expose the documentation which can't be extracted.

File: udata/core/user/api.py (hunk 6 above)
Comment (noirbizarre) [inline udata/core/user/api.py]: noted too :+1: (Probably was a given user before ;))

File: udata/templates/macros/issues.html (hunk 7 above)
Comment (noirbizarre) [inline udata/templates/macros/issues.html]: Yes, we can, but I find `__class__.__name__` explicit and consistent with the fact that we are avoiding necessary inheritance (one of the purpose of this PR).
I think a property hide the fact that it's simply the raw class name.
It's debatable, what do you think about it ?

File: udata/tests/test_issues.py (hunk 8 above)
Comment (noirbizarre) [inline udata/tests/test_issues.py]: `discussion` property is a legacy naming which might need renaming (in another PR) because now we have a `Discussion` model.

File: udata/core/followers/models.py [hunk 9]
```diff
@@ -15,16 +15,16 @@ class FollowQuerySet(db.BaseQuerySet):
     def following(self, user):
         return self(follower=user, until=None)
 
-    def followers(self, user):
-        return self(following=user, until=None)
+    def followers(self, followee):
```
Comment (vinyll) [inline udata/core/followers/models.py]: I don't mind having `user` here as it represent explicitely the type of object that is expected.

File: udata/core/organization/api.py [hunk 10]
```diff
@@ -296,7 +294,7 @@ def delete(self, org, user):
 
 @ns.route('/<id>/followers/', endpoint='organization_followers')
 class FollowOrgAPI(FollowAPI):
-    model = FollowOrg
+    model = Organization
```
Comment (vinyll) [inline udata/core/organization/api.py]: :+1:

File: udata/migrations/2015-12-09-generic-references.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Migrate disucssions and issues to DBRef/GenericReferenceField
```
Comment (vinyll) [inline udata/migrations/2015-12-09-generic-references.js]: _discussions_

File: udata/migrations/2015-12-09-generic-references.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+/**
+ * Maps inherited classes to theiru subject collection
```
Comment (vinyll) [inline udata/migrations/2015-12-09-generic-references.js]: _their_

File: udata/migrations/2015-12-09-generic-references.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+var CLASSES = {
+    'dataset': 'Dataset',
+    'reuse': 'Reuse',
+    'organization': 'Organization',
+    'user': 'User'
+};
+
+/**
+ * Produce a (wweird) GenericReferenceField from a class and an ID
```
Comment (vinyll) [inline udata/migrations/2015-12-09-generic-references.js]: _weird_ or is it on purpose?

File: udata/tests/api/test_model_resolution.py [hunk 14]
```diff
@@ -1,30 +0,0 @@
...
-        with self.assertRaises(ValueError):
-            api.resolve_model('UDataMongoEngine')
-
-    def test_raise_if_none(self):
-        with self.assertRaises(ValueError):
-            api.resolve_model(None)
-
-    def test_raise_if_missing_class_entry(self):
-        with self.assertRaises(ValueError):
-            api.resolve_model({'field': 'value'})
```
Comment (vinyll) [inline udata/tests/api/test_model_resolution.py]: Why is this test file deleted?

File: udata/tests/forms/test_reference_field.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+    def test_with_invalid_id(self):
+        Nested.objects.create(name=faker.name())
+        form = WithReferenceForm.from_json({'nested': 'bad'})
+
+        form.validate()
+        self.assertIn('nested', form.errors)
+        self.assertEqual(len(form.errors['nested']), 1)
+
+
+class Nested2(db.Document):
```
Comment (vinyll) [inline udata/tests/forms/test_reference_field.py]: "_Nested2_"? It would be worth having a short pydoc here. Same for `Nested3`.

File: js/components/dataset/card.vue [hunk 16]
```diff
@@ -92,7 +92,11 @@ import DatasetFilters from 'components/dataset/filters';
 import placeholders from 'helpers/placeholders';
 import config from 'config';
 
-const MASK = 'id,title,description,spatial{zones,granularity},metrics,organization,frequency,temporal_coverage';
+const MASK = [
```
Comment (vinyll) [inline js/components/dataset/card.vue]: As we discussed – and this is subjective – I know that `mask` actually refers to `X-Fields`in the API meta. I would therefore have called it _XFIELDS_ or _FIELDS_ to make it more uniform.
An extra opinion would be welcome.

File: js/views/dataset.vue [hunk 17]
```diff
@@ -168,17 +173,17 @@ export default {
...
         dataset: require('components/dataset/details.vue'),
         quality: require('components/dataset/quality.vue'),
         chart: require('components/charts/widget.vue'),
         resources: require('components/dataset/resource/list.vue'),
         followers: require('components/follow/list.vue'),
         wmap: require('components/widgets/map.vue'),
-        issues: require('components/issues/list.vue'),
-        discussions: require('components/discussions/list.vue'),
-        community: require('components/dataset/communityresource/list.vue'),
+        CommunityList,
```
Comment (vinyll) [inline js/views/dataset.vue]: is this some magic import that converts CamelCase to dashed-name?

File: js/views/editorial.vue [hunk 18]
```diff
@@ -3,19 +3,21 @@
     <div class="row">
         <dataset-cards class="col-xs-12 col-md-6"
             :title="_('Featured datasets')"
-            :datasets="home_datasets.items">
+            :datasets="home_datasets.items"
+            :loading="home_datasets.loading">
         </dataset-cards>
-        <reuses class="col-xs-12 col-md-6"
+        <reuse-cards class="col-xs-12 col-md-6"
```
Comment (vinyll) [inline js/views/editorial.vue]: `reuse-list`?

Comment (vinyll) [inline js/views/editorial.vue]: `<reuse-card-list>`?

File: js/views/editorial.vue [hunk 19]
```diff
@@ -3,19 +3,21 @@
     <div class="row">
         <dataset-cards class="col-xs-12 col-md-6"
```
Comment (vinyll) [inline js/views/editorial.vue]: `<dataset-card-list>`?

File: js/views/site.vue [hunk 20]
```diff
@@ -18,7 +18,7 @@
         <reuse-list id="reuses" class="col-xs-12" :reuses="reuses"></reuse-list>
     </div>
     <div class="row">
-        <organizations id="organizations" class="col-xs-12" :organizations="organizations"></organizations>
+        <org-list id="organizations" class="col-xs-12" :organizations="organizations"></org-list>
```
Comment (vinyll) [inline js/views/site.vue]: `<organization-list>`?

File: js/views/home.vue [hunk 21]
```diff
@@ -40,7 +41,8 @@ export default {
             metrics: new MyMetrics(),
             reuses: new PageList({
                 ns: 'me',
-                fetch: 'my_org_reuses'
+                fetch: 'my_org_reuses',
```
Comment (vinyll) [inline js/views/home.vue]: _my_organization_reuses_?

File: js/views/search.vue [hunk 22]
```diff
@@ -51,15 +52,16 @@ export default {
             datasets: new PageList({
                 ns: 'me',
                 fetch: 'my_org_datasets',
```
Comment (vinyll) [inline js/views/search.vue]: _my_organization_datasets_?

File: js/views/search.vue [hunk 23]
```diff
@@ -51,15 +52,16 @@ export default {
...
+                mask: DatasetList.MASK
             }),
             communities: new PageList({
                 ns: 'me',
                 fetch: 'my_org_community_resources'
             }),
             reuses: new PageList({
                 ns: 'me',
-                fetch: 'my_org_reuses'
+                fetch: 'my_org_reuses',
```
Comment (vinyll) [inline js/views/search.vue]: _my_organization_reuses_?

File: js/views/editorial.vue [hunk 24]
```diff
@@ -1,10 +1,10 @@
 <template>
 <layout :title="_('Editorial')">
     <div class="row">
-        <datasets class="col-xs-12 col-md-6"
+        <dataset-cards class="col-xs-12 col-md-6"
```
Comment (vinyll) [inline js/views/editorial.vue]: `<dataset-card-list>`?

File: js/views/me.vue [hunk 25]
```diff
@@ -7,7 +7,7 @@
     </div>
 
     <div class="row">
-        <datasets class="col-xs-12" :datasets="datasets"></datasets>
+        <datasets-list class="col-xs-12" :datasets="datasets"></datasets-list>
```
Comment (vinyll) [inline js/views/me.vue]: `<dataset-list>`
We have `<reuse-list>` and others with no _s_

File: js/views/search.vue [hunk 26]
```diff
@@ -49,7 +50,8 @@ export default {
         return {
             datasets: new PageList({
                 ns: 'me',
-                fetch: 'my_org_datasets'
+                fetch: 'my_org_datasets',
```
Comment (vinyll) [inline js/views/search.vue]: _my_organization_datasets_?

File: udata/templates/macros/issues.html (hunk 7 above)
Comment (vinyll) [inline udata/templates/macros/issues.html]: I really don't mind using `__class__.__name__` anywhere in Python but I don't think it's appropriate in a template.

**Comment by vinyll** [comment]
Massively impressive PR. :open_mouth:

File: udata/templates/macros/issues.html (hunk 7 above)
Comment (noirbizarre) [inline udata/templates/macros/issues.html]: But here, to be able to do that properly, the property needs to be declared on each model because right now we don't have a common ancestor.
I agree with you, template should stay simple, but in this specific case, the main point is that it's a standard Python object (a class instance) instead of a specific inherited `*Discussion`.

File: udata/core/followers/models.py (hunk 9 above)
Comment (noirbizarre) [inline udata/core/followers/models.py]: The debate is open: I can switch back to user. It depends on what you find clearer.
The vocabulary here is a little tricky: `user`/`followee`/`follower`/`following`.
Until then, I switched back to user but I think the `follow` module need a little bit of naming coherence in another PR.

File: udata/tests/api/test_model_resolution.py (hunk 14 above)
Comment (noirbizarre) [inline udata/tests/api/test_model_resolution.py]: Because this part moved on models/db and so the tests moved into `test_models`.

File: js/views/dataset.vue (hunk 17 above)
Comment (noirbizarre) [inline js/views/dataset.vue]: yes, it is :) But it's very well documented:
- https://vuejs.org/guide/components.html#Assets-Naming-Convention
- https://vuejs.org/guide/components.html#camelCase-vs-kebab-case

This is mainly du to the fact that on one side, html is case insensitive and recommend using dash into WebComponents name and properties and on the other side, dash is not legit in JavaScript naming and CamelCase is the convention.

File: js/views/home.vue (hunk 21 above)
Comment (noirbizarre) [inline js/views/home.vue]: Yes, maybe in another PR on `org` vs `organization` naming convention (this one require to change on the API side too)

File: js/views/search.vue (hunk 23 above)
Comment (noirbizarre) [inline js/views/search.vue]: same as above

File: js/views/search.vue (hunk 26 above)
Comment (noirbizarre) [inline js/views/search.vue]: Same as above

File: js/components/reuse/card.vue [hunk 27]
```diff
@@ -64,7 +64,18 @@ import placeholders from 'helpers/placeholders';
 import moment from 'moment';
 import config from 'config';
 
+const MASK = [
```
Comment (davidbgk) [inline js/components/reuse/card.vue]: Consistency? ;)

File: js/views/reuse.vue [hunk 28]
```diff
@@ -38,8 +38,14 @@ import Metrics from 'models/metrics';
 import Vue from 'vue';
 import Issues from 'models/issues';
 import Discussions from 'models/discussions';
+// Widgets
+import DatasetCards from 'components/dataset/card-list.vue';
+import DiscussionList from 'components/discussions/list.vue';
+import IssueList from 'components/issues/list.vue';
 import Layout from 'components/layout.vue';
 
+const MASK = `datasets{${DatasetCards.MASK}},*`;
```
Comment (davidbgk) [inline js/views/reuse.vue]: Do we really need all fields for reuses?

File: udata/api/__init__.py [hunk 29]
```diff
@@ -34,6 +34,23 @@
...
+FACETS_TYPES = {
+    search.fields.BoolFacet: inputs.boolean
+}
+
+
+def description():
+    '''Load description from markdown file'''
+    if 'short' in request.args:
+        return ''
+    global DESCRIPTION
```
Comment (davidbgk) [inline udata/api/__init__.py]: Ugh, can't we load that once and for all on server launch?

File: udata/api/description.md [hunk 30]
```diff
@@ -0,0 +1,42 @@
+## {{ _('Authentication') }}
```
Comment (davidbgk) [inline udata/api/description.md]: That's weird to have that here, maybe it should be part/included from the docs?

File: udata/core/dataset/api_fields.py [hunk 31]
```diff
@@ -101,6 +101,14 @@
 community_resource_page_fields = api.model(
     'CommunityResourcePage', fields.pager(community_resource_fields))
 
+#: Default mask to make it lightweight by default
+DEFAULT_MASK = ','.join((
```
Comment (davidbgk) [inline udata/core/dataset/api_fields.py]: It seconds my thoughts on consistency for masks, we should stick to lists on the js side too.

File: udata/migrations/2015-12-09-generic-references.js [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+            count++;
+        }
+    });
+    print('Processed ' + count + ' documents in db.' + collection);
+}
+
+process('discussion', 'subject');
+process('issue', 'subject');
+process('follow', 'following');
+// process('activity', 'related_to');
```
Comment (davidbgk) [inline udata/migrations/2015-12-09-generic-references.js]: Why not activity anymore?

File: udata/core/dataset/models.py [hunk 33]
```diff
@@ -324,7 +322,11 @@ def quality(self):
             * description length
             * and so on
         """
+        from udata.models import Discussion
```
Comment (davidbgk) [inline udata/core/dataset/models.py]: If it's a circular import, add a comment.

File: js/components/reuse/card.vue (hunk 27 above)
Comment (noirbizarre) [inline js/components/reuse/card.vue]: Different day, different mood, different solution :)
Which one do you prefer: single line or joined array ?

File: js/views/reuse.vue (hunk 28 above)
Comment (noirbizarre) [inline js/views/reuse.vue]: I've chosen the shorter version: required dataset fields and all reuse fields (not so many bu most of them used on the page). The biggest data was the datasets payload (and all cards are now fetch once with only the required fields vs one call by card before with all fields)

File: js/components/reuse/card.vue (hunk 27 above)
Comment (davidbgk) [inline js/components/reuse/card.vue]: Arrays, joined when necessary, not during the definition otherwise we'll have to split if again if we need to merge/modify it later.

File: udata/api/__init__.py (hunk 29 above)
Comment (noirbizarre) [inline udata/api/__init__.py]: Not meant to be in the PR. I'll extract it. It was an experiment having as goal:
- all the documentation in the swagger (meant for it)
- description into a single templatable and translatable markdown file (one by langue, we translate the whole file)
- description serialized only when needed (not in admin as example)

File: udata/core/dataset/api_fields.py (hunk 31 above)
Comment (noirbizarre) [inline udata/core/dataset/api_fields.py]: In this PR, the position was:
- mask as string by default
- as array only when needed or better (here only to be able to have a line under 80 char)

On the client side I wanted to avoir array joining every time. But using joined array by default is fine too me too.

File: udata/migrations/2015-12-09-generic-references.js (hunk 32 above)
Comment (noirbizarre) [inline udata/migrations/2015-12-09-generic-references.js]: Because right now, inheritance is an entire part of the Activity management algorythm (whereas on other classes it's only there to type the subject which is exactly what `GenericReferenceField` do) and to remove it it needs rewriting the activity handling.
It might be necessary but I think it was too much for this PR and deserve a separate PR.

File: udata/core/dataset/models.py (hunk 33 above)
Comment (noirbizarre) [inline udata/core/dataset/models.py]: Yes, it's for avoid them in futur.
I can't help wondering it we are not missing a layer here. Having so much computation in a property might need action action layer.

Comment (davidbgk) [inline udata/core/dataset/models.py]: The computation has been moved to `compute_quality_score` and `check_availability`, we are just assembling the dict here. But I'm open to better suggestions, maybe it should be done completely asynchronously, or even externalized…

File: js/models/mask.js [hunk 34]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Serialize a mask as String for header
+ * @param  {String|Array} mask The mask to serialize
+ * @return {String}      A Serialized mask ready to be used into HTTP header
+ */
+export function serialize(mask) {
```
Comment (davidbgk) [inline js/models/mask.js]: 👍

File: udata/api/__init__.py [hunk 35]
```diff
@@ -34,6 +33,11 @@
 
 DEFAULT_PAGE_SIZE = 50
 HEADER_API_KEY = 'X-API-KEY'
+DESCRIPTION = None
```
Comment (davidbgk) [inline udata/api/__init__.py]: Still useful?

File: udata/migrations/2015-12-09-generic-references.js (hunk 32 above)
Comment (davidbgk) [inline udata/migrations/2015-12-09-generic-references.js]: In that case remove that line and related `TYPES` definitions?

**Comment by davidbgk** [comment]
r+wc
